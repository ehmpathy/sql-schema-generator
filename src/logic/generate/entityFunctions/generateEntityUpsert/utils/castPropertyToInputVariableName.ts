/*
  takes a property name and converts it into what we would define the upsert input name to be
*/
export const castPropertyToInputVariableName = ({ name }: { name: string }) => `in_${name}`;
