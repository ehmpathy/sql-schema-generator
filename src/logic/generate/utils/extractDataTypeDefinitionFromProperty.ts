import { Property } from '../../../domain';

// https://www.postgresql.org/docs/10/datatype.html
export const extractDataTypeDefinitionFromProperty = ({
  property,
}: {
  property: Property;
}) => {
  let typeModifier = '';
  if (property.type.precision) {
    typeModifier = `(${property.type.precision})`;
  }
  if (property.type.precision && property.type.scale) {
    typeModifier = `(${property.type.precision}, ${property.type.scale})`;
  }
  if (property.array) {
    typeModifier += '[]';
  }
  return `${property.type.name}${typeModifier}`;
};
