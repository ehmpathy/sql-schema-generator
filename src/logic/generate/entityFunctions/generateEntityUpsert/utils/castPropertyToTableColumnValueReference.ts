import { Property } from '../../../../../domain';
import { castPropertyToInputVariableName } from './castPropertyToInputVariableName';

/*
  defines the input value reference for the main tables (static and version)
    - purpose: abstract away the fact that we have to "hash" the input for array variables

  NOTE: this is not used by the mapping tables - as they have a special way of parsing the input variable value
    - i.e., they parse and loop through it
    - not to mention they can not use the hash value
*/
export const castPropertyToTableColumnValueReference = ({
  name,
  definition,
}: {
  name: string;
  definition: Property;
}) => {
  const inputVariableName = castPropertyToInputVariableName({ name });

  // if it is an array, then hash the input value into a binary value
  if (definition.array)
    return `digest(array_to_string(${inputVariableName}, ',', '__NULL__'), 'sha256')`;

  // if not array, then return the input variable reference directly
  return inputVariableName;
};
