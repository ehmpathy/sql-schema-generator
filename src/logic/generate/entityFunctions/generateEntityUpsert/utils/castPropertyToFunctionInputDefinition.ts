import { Property } from '../../../../../domain';
import { extractDataTypeDefinitionFromProperty } from '../../../utils/extractDataTypeDefinitionFromProperty';
import { castPropertyToInputVariableName } from './castPropertyToInputVariableName';

/*
  NOTE: special logic added here to treat array properties.
    For array properties, the input type is different from that defined on the property - due to being an "array" of those properties
*/

export const castPropertyToFunctionInputDefinition = ({ name, definition }: { name: string; definition: Property }) => {
  const inputVariableName = castPropertyToInputVariableName({ name });
  const inputVariableType = extractDataTypeDefinitionFromProperty({ property: definition });
  return `${inputVariableName} ${inputVariableType}`;
};
