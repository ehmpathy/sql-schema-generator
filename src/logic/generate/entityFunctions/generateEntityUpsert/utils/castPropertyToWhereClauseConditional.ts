import { Property } from '../../../../../types';
import { castPropertyToColumnName } from '../../../utils/castPropertyToColumnName';
import { castPropertyToTableColumnValueReference } from './castPropertyToTableColumnValueReference';

export const castPropertyToWhereClauseConditional = ({ name, definition }: { name: string; definition: Property }) => {
  const columnName = castPropertyToColumnName({ name, definition }); // no change
  const columnValueReference = castPropertyToTableColumnValueReference({ name, definition });
  return [
    `AND (${columnName} = BINARY ${columnValueReference}`, // use binary comparison to ensure case sensitivity (even though column may be case sensitive, mysql will still "help us out" by defaulting to case insensitive)
    definition.nullable ? ` OR (${columnName} IS null AND ${columnValueReference} IS null)` : '', // NULL != NULL, so special check if field is nullable
    ')',
  ].join('');
};
