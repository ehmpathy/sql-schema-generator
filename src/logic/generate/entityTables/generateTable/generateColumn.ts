import { DataTypeName, Property } from '../../../../types';
import { extractMysqlTypeDefinitionFromProperty } from '../../utils/extractMysqlTypeDefinitionFromProperty';

/*
  // TODO: generalize w/ adapter pattern to other languages
*/
const textTypes = [
  DataTypeName.CHAR,
  DataTypeName.VARCHAR,
  DataTypeName.TINYTEXT,
  DataTypeName.TEXT,
  DataTypeName.MEDIUMTEXT,
  DataTypeName.LONGTEXT,
  DataTypeName.ENUM,
];
export const generateColumn = ({ columnName, property }: { columnName: string, property: Property }) => {

  // define default over-rides
  let defaultNullRelation = (property.nullable) ? 'DEFAULT NULL' : ''; // if nullable, then the "default" default is "null"; else, nothing
  if (property.type.name === DataTypeName.TEXT) defaultNullRelation = ''; // if its varchar, then even if nullable don't show default as null; // TODO: discover why SHOW CREATE has this exception

  // define the full column definition
  const modifiers = [
    `\`${columnName}\``,

    extractMysqlTypeDefinitionFromProperty({ property }), // e.g., property => bigint(20)

    (textTypes.includes(property.type.name)) ? 'COLLATE utf8_bin' : '', // coalate if its a string type; collate to Case Sensitive by default

    (property.nullable) ? '' : 'NOT NULL', // if not nullable, then NOT NULL

    (property.default) ? `DEFAULT ${property.default}` : defaultNullRelation, // if default is set, use it; otherwise, no default

    (property.comment) ? `COMMENT '${property.comment}'` : '', // empty string if comment not defined

    (columnName === 'id') ? 'AUTO_INCREMENT' : '',
  ];
  return modifiers
    .filter(modifier => !!modifier) // filter out null
    .join(' '); // merge into column
};
