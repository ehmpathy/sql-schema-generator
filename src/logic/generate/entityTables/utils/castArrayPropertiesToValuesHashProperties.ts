import { Property } from '../../../../types';
import { prop } from '../../../define';
import { castPropertyToColumnName } from '../../utils/castPropertyToColumnName';

/*
  we store the hash of the array values on the tables themselves, for performance and simplicity in comparison queries
    even though the data is mastered through mapping tables

  this gives us a utility to "create a values hash property for each array property"
*/
export const castArrayPropertiesToValuesHashProperties = ({
  properties,
}: {
  properties: { [index: string]: Property };
}) => {
  const castedProperties: { [index: string]: Property } = {};
  Object.entries(properties).forEach(([name, definition]) => {
    if (!definition.array) {
      throw new Error('error - non array property was asked to have been casted into values hash property');
    }
    const columnName = castPropertyToColumnName({ name, definition });
    castedProperties[columnName] = prop.CHAR(64); // SHA2(in_string, 256) produces a 64 char string
  });
  return castedProperties;
};
