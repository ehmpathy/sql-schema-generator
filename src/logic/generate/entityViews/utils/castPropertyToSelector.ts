import { Property } from '../../../../types';
import { defineMappingTableKeysForEntityProperty } from '../../utils/defineMappingTableKeysForEntityProperty';

export const castPropertyToSelector = ({
  entityName,
  name,
  definition,
}: {
  entityName: string;
  name: string;
  definition: Property;
}) => {
  // if property is an array, the selector should CONCAT_WS from the mapping table
  if (definition.array) {
    const mappingTableKeys = defineMappingTableKeysForEntityProperty({ entityName, propertyDefinition: definition });
    const entityReferenceTableNameAlias = definition.updatable ? 'v' : 's';
    const arrayValueSelector = `${mappingTableKeys.tableName}.${mappingTableKeys.mappedEntityReferenceColumnName}`;
    const arrayIndexSelector = `${mappingTableKeys.tableName}.${mappingTableKeys.arrayOrderIndexColumnName}`;
    const entityReferenceSelector = `${mappingTableKeys.tableName}.${mappingTableKeys.entityReferenceColumnName}`;
    const entityReferenceId = `${entityReferenceTableNameAlias}.id`;
    const mappingTableName = mappingTableKeys.tableName;
    return `
(
  SELECT GROUP_CONCAT(${arrayValueSelector} ORDER BY ${arrayIndexSelector} ASC SEPARATOR ',')
  FROM ${mappingTableName} WHERE ${entityReferenceSelector} = ${entityReferenceId}
) as ${name}
    `.trim();
  }

  // otherwise, if its an updatable property, its on the version table
  if (definition.updatable) return `v.${name}`;

  // and if its not updatable and not array, then its on the static table
  return `s.${name}`;
};
