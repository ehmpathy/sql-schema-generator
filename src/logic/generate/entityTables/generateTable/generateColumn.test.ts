import { DataType, DataTypeName, Property } from '../../../../domain';
import { extractDataTypeDefinitionFromProperty } from '../../utils/extractDataTypeDefinitionFromProperty';
import { generateColumn } from './generateColumn';

jest.mock('../../utils/extractDataTypeDefinitionFromProperty');
const extractDataTypeDefinitionFromPropertyMock = extractDataTypeDefinitionFromProperty as jest.Mock;
extractDataTypeDefinitionFromPropertyMock.mockReturnValue('__MOCKED_TYPE_DEFINITION__');

describe('generateColumn', () => {
  beforeEach(() => jest.clearAllMocks());
  it('should define column name accurately', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.INT,
      }),
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain('user_id');
    expect(sql).toMatchSnapshot(); // to log an example, not to actualy test logic
  });
  it('should define data_type definition accurately', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.INT,
      }),
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(extractDataTypeDefinitionFromPropertyMock.mock.calls.length).toEqual(1);
    expect(extractDataTypeDefinitionFromPropertyMock.mock.calls[0][0]).toMatchObject({ property });
    expect(sql).toContain('__MOCKED_TYPE_DEFINITION__');
  });
  it('should specify NOT NULL by default', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.INT,
      }),
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain('NOT NULL');
    expect(sql).toMatchSnapshot(); // to log an example, not to actualy test logic
  });
  it('should specify NULL if nullable', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.INT,
      }),
      nullable: true,
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain('NULL');
    expect(sql).not.toContain('NOT NULL');
    expect(sql).toMatchSnapshot(); // to log an example, not to actualy test logic
  });
  it('should be able to specify a default value', () => {
    const property = new Property({
      type: new DataType({
        name: DataTypeName.TIMESTAMPTZ,
      }),
      default: 'now()',
    });
    const sql = generateColumn({ columnName: 'user_id', property });
    expect(sql).toContain('DEFAULT now()');
    expect(sql).toMatchSnapshot(); // to log an example, not to actualy test logic
  });
});
