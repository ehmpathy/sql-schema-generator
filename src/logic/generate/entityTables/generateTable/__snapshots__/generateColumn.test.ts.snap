// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateColumn should be able to specify a default value 1`] = `"user_id __MOCKED_TYPE_DEFINITION__ NOT NULL DEFAULT now()"`;

exports[`generateColumn should define column name accurately 1`] = `"user_id __MOCKED_TYPE_DEFINITION__ NOT NULL"`;

exports[`generateColumn should specify NOT NULL by default 1`] = `"user_id __MOCKED_TYPE_DEFINITION__ NOT NULL"`;

exports[`generateColumn should specify NULL if nullable 1`] = `"user_id __MOCKED_TYPE_DEFINITION__ NULL"`;
